js/ui/components/codingGameService.js

```diff
@@ -1,52 +1,152 @@
 
 const Gio = imports.gi.Gio;
+const GLib = imports.gi.GLib;
 const Lang = imports.lang;
 const Meta = imports.gi.Meta;
-const Showmehow = imports.gi.Showmehow;
+const CodingGameService = imports.gi.CodingGameService;
 
 const CodingChatboxTextService = new Lang.Class({
     Name: 'CodingChatboxTextService',
 
     enable: function() {
         // Connect to the service
         try {
-            this._service = Showmehow.ServiceProxy.new_for_bus_sync(
+            this._service = CodingGameService.CodingGameServiceProxy.new_for_bus_sync(
                 Gio.BusType.SESSION, 0,
-                'com.endlessm.Showmehow.Service',
-                '/com/endlessm/Showmehow/Service', null);
+                'com.endlessm.CodingGameService.Service',
+                '/com/endlessm/CodingGameService/Service', null);
         } catch (e) {
-            logError(e, 'Error occurred in creating ShowmehowServiceProxy');
+            logError(e, 'Error occurred in creating CodingGameServiceProxy');
             return;
         }
 
-        // It doesn't seem possible to return an array of strings here, looks like it has to be an array
-        // of tuples (which contain strings).
-        this._listenForLessonsId =
-            this._service.connect('listening-for-lesson-events', Lang.bind(this, function(proxy, interestingEvents) {
-                let events = interestingEvents.deep_unpack().map(function(i) {
-                    return i[0];
-                });
+        let eventHandlers = {
+            'move-window': Lang.bind(this, this._waitForNextWindowMove),
+            'stop-moving-windows': Lang.bind(this, this._waitForNoWindowsToMove)
+        };
 
-                if (events.indexOf('window-moved') !== -1)
-                    this._waitForNextWindowMove();
+        this._signalHandlers = {};
+        this._listenForEventsId =
+            this._service.connect('listen-for-event', Lang.bind(this, function(proxy, name) {
+                if (eventHandlers[name]) {
+                    this._disconnectHandlersFor(name);
+                    eventHandlers[name]();
+                }
             }));
+        this._stopListeningForEventsId =
+            this._service.connect('stop-listening-for', Lang.bind(this, function(proxy, name) {
+                // In this case, just disconnect any known event handlers, don't
+                // need any custom logic here
+                this._disconnectHandlersFor(name);
+            }));
+
+        this._service.call_currently_listening_for_events(null, Lang.bind(this, function(source, result) {
+            let success, returnValue;
+
+            try {
+                [success, returnValue] = this._service.call_currently_listening_for_events_finish(result);
+            } catch (e) {
+                logError(e);
+                return;
+            }
+
+            returnValue.deep_unpack().map(Lang.bind(this, function(name) {
+                if (eventHandlers[name]) {
```

Comment:
Please factor out this block in a separate function, since it's also repeated above